### PR TITLE
[tt-train] Add custom Python operations support for TTML autograd

### PR DESCRIPTION
**Pull request by ayerofieiev-tt**
## Summary

This PR adds support for custom Python operations in TTML autograd, following PyTorch's `torch.autograd.Function` pattern. Users can now define custom operations with forward and backward passes directly in Python, integrating seamlessly with TTML's existing C++ autograd system.

### Two Usage Patterns

1. **Compose with existing autograd ops** - When forward uses TTML autograd operations (like `+`, `-`, `*`, or `ttml.ops.*`), backward is automatic:
   ```python
   class AddAndScale(Function):
       @staticmethod
       def forward(ctx, x, y, scale):
           return (x + y) * scale  # Graph builds automatically
       # No backward needed!
   ```

2. **Build from ttnn primitives** - When using low-level ttnn operations, implement backward explicitly:
   ```python
   class Scale(Function):
       @staticmethod
       def forward(ctx, input, scale_factor):
           ctx.save_for_backward(input)
           ctx.scale_factor = scale_factor
           return ttnn.multiply(input.get_value(), scale_factor)
       
       @staticmethod
       def backward(ctx, grad_output):
           return ttnn.multiply(grad_output, ctx.scale_factor)
   ```

### Core Components

- **`Function` base class**: PyTorch-like API for defining custom operations
- **`FunctionContext`**: Context object for saving tensors and values needed in backward pass
- **`get_links()`**: Utility function to extract computation graph links from tensors
- **Auto-wrapping**: ttnn tensors returned from forward are automatically wrapped as ttml autograd tensors
- **Gradient validation**: Validates that backward returns correct number of gradients

### C++ Bindings Updates

- Added `get_gradient_mode()` method to AutoContext
- Added `create_tensor()` module-level functions for creating autograd tensors
- Improved `add_backward_node()` binding to handle empty lists correctly
- Exposed `NodeId` class to Python

### Testing

Comprehensive test suite with 575+ lines covering:
- FunctionContext save/retrieve functionality
- get_links() utility with various edge cases
- Custom operations with device (integration tests)
- Custom operations without device (unit tests)
- Gradient mode handling (enabled/disabled)
- Multiple inputs and outputs
- Mixed tensor/scalar inputs
- Auto-wrapping of ttnn tensors
- Gradient accumulation and validation
- Both usage patterns (compose vs. ttnn primitives)

## Files Changed

- `tt-train/docs/CUSTOM_AUTOGRAD_FUNCTIONS.md` - Complete tutorial with examples (238 lines)
- `tt-train/sources/ttml/nanobind/nb_autograd.cpp` - C++ bindings updates
- `tt-train/sources/ttml/ttml/autograd/__init__.py` - New autograd package init (73 lines)
- `tt-train/sources/ttml/ttml/autograd/function.py` - Core Function implementation (309 lines)
- `tt-train/tests/python/conftest.py` - Pytest configuration with device markers (36 lines)
- `tt-train/tests/python/test_custom_operations.py` - Comprehensive test suite (575 lines)

**Total**: 6 files changed, 1,254 insertions(+), 1 deletion(-)

## API Comparison with PyTorch

This implementation closely follows PyTorch's `torch.autograd.Function` API for familiarity:

| Feature | PyTorch | TTML | Status |
|---------|---------|------|--------|
| **API Pattern** | `class MyOp(torch.autograd.Function)` | `class MyOp(ttml.autograd.Function)` | ✅ Identical |
| **Forward Method** | `@staticmethod def forward(ctx, *inputs)` | `@staticmethod def forward(ctx, *inputs)` | ✅ Identical |
| **Backward Method** | `@staticmethod def backward(ctx, *grad_outputs)` | `@staticmethod def backward(ctx, *grad_outputs)` | ✅ Identical |
| **Apply Method** | `MyOp.apply(*inputs)` | `MyOp.apply(*inputs)` | ✅ Identical |
| **Context API** | `ctx.save_for_backward()`, `ctx.saved_tensors` | `ctx.save_for_backward()`, `ctx.saved_tensors` | ✅ Identical |
| **Arbitrary Attributes** | `ctx.scale = 2.0` | `ctx.scale = 2.0` | ✅ Identical |
| **Gradient Returns** | Returns gradients (not in-place) | Returns gradients (not in-place) | ✅ Identical |
| **Multiple Outputs** | Returns tuple of tensors | Returns tuple of tensors | ✅ Supported |
| **None Handling** | Returns None for non-differentiable inputs | Returns None for non-differentiable inputs | ✅ Supported |

### Key Differences

- **Tensor Creation**: Use `ttml.autograd.create_tensor(value, requires_grad)` or return ttnn tensors (auto-wrapped)
- **Operations**: Use `ttnn.multiply()`, `ttnn.add()`, etc. instead of `torch.mul()`, `torch.add()`
- **Gradient Types**: Backward receives raw `tt::tt_metal::Tensor` gradients (C++ backend)
- **Tensor Access**: Use `tensor.get_value()` to access underlying tensor for ttnn operations

## Example Usage

```python
import ttml
import ttnn
import numpy as np

# Pattern 1: Compose with autograd ops (backward automatic)
class AddAndScale(Function):
    @staticmethod
    def forward(ctx, x, y, scale):
        return (x + y) * scale  # Uses ttml tensor ops

# Pattern 2: Build from ttnn primitives (explicit backward)
class Scale(Function):
    @staticmethod
    def forward(ctx, input, scale_factor):
        ctx.save_for_backward(input)
        ctx.scale_factor = scale_factor
        # Return ttnn tensor - auto-wrapped
        return ttnn.multiply(input.get_value(), scale_factor)
    
    @staticmethod
    def backward(ctx, grad_output):
        # grad_output is raw tt::tt_metal::Tensor
        return ttnn.multiply(grad_output, ctx.scale_factor)

# Usage
x = ttml.autograd.Tensor.from_numpy(np.array([[1, 2], [3, 4]], dtype=np.float32))
y = Scale.apply(x, 2.0)
y.backward()  # Gradients flow to x
```

## Documentation

Complete tutorial added in `CUSTOM_AUTOGRAD_FUNCTIONS.md` covering:
- Both usage patterns with examples
- Multiple inputs/outputs handling
- Implicit tensor conversion (auto-wrapping)
- Gradient validation
- Complete Legendre polynomial example (similar to PyTorch tutorial)

## Testing

All tests pass, including:
- ✅ FunctionContext functionality
- ✅ get_links() utility
- ✅ Forward/backward passes
- ✅ Gradient accumulation
- ✅ Multiple inputs/outputs
- ✅ Gradient mode handling
- ✅ Auto-wrapping of ttnn tensors
- ✅ Both usage patterns

The implementation is production-ready and follows PyTorch conventions for easy migration.

### CODE REVIEW

File: tt-train/sources/ttml/ttml/autograd/function.py [hunk 1]
```diff
@@ -0,0 +1,261 @@
...
+            # Register the backward node
+            node_id = auto_context.add_backward_node(backward_closure, links)
+
+            # Set the node on all output tensors
+            if node_id is not None:
+                for output in outputs_tuple:
+                    if output is not None and hasattr(output, "set_node"):
+                        output.set_node(node_id)
+
+        return outputs if not single_output else outputs
```
[the remark this answers is not in the document]
Comment (mdragulaTT) [reply tt-train/sources/ttml/ttml/autograd/function.py]: Valid

File: tt-train/sources/ttml/ttml/autograd/function.py [hunk 2]
```diff
@@ -0,0 +1,264 @@
...
+        List of NodeId objects from tensors that have nodes set.
+    """
+    links = []
+    for tensor in tensors:
+        if tensor is None:
+            continue
+        node = tensor.get_node()
+        if node is not None:
+            links.append(node)
+    return links
```
Comment (mdragulaTT) [inline tt-train/sources/ttml/ttml/autograd/function.py:104]: maybe sth like
```
return [
        tensor.get_node()
        for tensor in tensors
        if tensor is not None and tensor.get_node() is not None
    ]
```
?

**Review by mdragulaTT — COMMENTED** [review]
Overall LGTM, just address the Copilot comments or discard, but overall is good

File: tt-train/sources/ttml/ttml/autograd/function.py [hunk 3]
```diff
@@ -0,0 +1,264 @@
...
+
+
+def _get_cpp_bindings():
+    """Lazily import C++ bindings to avoid circular import issues."""
+    global _cpp_bindings
+    if _cpp_bindings is None:
+        from ttml import _ttml
+
+        _cpp_bindings = _ttml.autograd
+    return _cpp_bindings
```
Comment (philei-tt) [inline tt-train/sources/ttml/ttml/autograd/function.py]: This is weird

File: tt-train/sources/ttml/ttml/autograd/function.py [hunk 4]
```diff
@@ -0,0 +1,264 @@
...
+    def __init__(self):
+        self._saved_tensors: List[Any] = []
+
+    def save_for_backward(self, *tensors) -> None:
+        """Save tensors for use in backward pass.
+
+        Args:
+            *tensors: Tensors to save. Can include None values which will be preserved.
+        """
+        self._saved_tensors = list(tensors)
```
Comment (philei-tt) [inline tt-train/sources/ttml/ttml/autograd/function.py:47]: Honestly ugly, but kind of make sense. I didn't realize torch works like that. 

Alternative would be something more similar to cpp: 
```python
class Scale(torch.autograd.Function):
    def __call__(ctx, input, scale_factor):
        ctx.scale_factor = scale_factor
        output = input * scale_factor
        def backward(grad_out):
            return grad_output * ctx.scale_factor
        return output, backward
```

@ayerofieiev-tt @athompsonTT  wdyt? Should we stick to api similar to torch?

File: tt-train/sources/ttml/ttml/autograd/function.py (hunk 2 above)
Comment (philei-tt) [reply tt-train/sources/ttml/ttml/autograd/function.py:104]: I guess this is more "pythonic", but imo what's now is more readable

File: tt-train/sources/ttml/ttml/autograd/__init__.py [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+    import ttnn
+
+    # Define custom operation
+    class Scale(ttml.autograd.Function):
+        @staticmethod
+        def forward(ctx, input, scale):
+            ctx.save_for_backward(input)
+            ctx.scale = scale  # Save scalars as attributes
+            output = ttnn.multiply(input.get_value(), scale)
+            return ttml.autograd.create_tensor(output, requires_grad=True)
```
Comment (philei-tt) [inline tt-train/sources/ttml/ttml/autograd/__init__.py:24]: I don't like passing requires_grad flag, this is unnecessary complexity. I think we should allow user to return both ttnn and ttml tensor, and then based on type `apply` will convert to ttml tensor if ttnn is returned.

File: tt-train/sources/ttml/ttml/autograd/function.py [hunk 6]
```diff
@@ -0,0 +1,264 @@
...
+        # Collect input tensors (objects with get_requires_grad method)
+        input_tensors = [inp for inp in inputs if hasattr(inp, "get_requires_grad")]
+
+        # Check if any input requires gradients
+        any_requires_grad = any(t.get_requires_grad() for t in input_tensors)
+
+        # Only set up backward if we have inputs that require gradients
+        if any_requires_grad:
+            # Extract links from input tensors (for graph dependencies)
+            links = get_links(input_tensors)
```
Comment (philei-tt) [inline tt-train/sources/ttml/ttml/autograd/function.py]: We should add node unconditionally. This optimization is going to be performed in cpp and will be added as  a part of lora pr

File: tt-train/sources/ttml/ttml/autograd/function.py [hunk 7]
```diff
@@ -0,0 +1,264 @@
...
+                        # Get gradients from output tensors
+                        grad_outputs = []
+                        for out in bwd_outputs:
+                            if out is not None and out.is_grad_initialized():
+                                grad_outputs.append(out.get_grad())
+                            else:
+                                raise RuntimeError(
+                                    f"Output tensor gradient not initialized in "
+                                    f"{bwd_cls.__name__}.backward()"
+                                )
```
Comment (philei-tt) [inline tt-train/sources/ttml/ttml/autograd/function.py]: I need to double check, but I think this should fail every time. gradients are initialized lazily when you call tensor.add_grad

File: tt-train/sources/ttml/ttml/autograd/function.py [hunk 8]
```diff
@@ -0,0 +1,264 @@
...
+                            grad_inputs = bwd_cls.backward(bwd_ctx, *grad_outputs)
+
+                        # Normalize to tuple
+                        if grad_inputs is None:
+                            grad_inputs = (None,) * len(bwd_inputs)
+                        elif not isinstance(grad_inputs, tuple):
+                            grad_inputs = (grad_inputs,)
+
+                        # Accumulate gradients to input tensors
+                        for tensor, grad in zip(bwd_inputs, grad_inputs):
```
Comment (philei-tt) [inline tt-train/sources/ttml/ttml/autograd/function.py]: This assumption that we should return the same number of tensors in backward as number of inputs, and the order should be matched is a bit fragile imo. At least we should add a very descriptive error message, and make sure this is well explained in docs

**Review by philei-tt — CHANGES_REQUESTED** [review]

File: tt-train/sources/ttml/ttml/autograd/function.py (hunk 4 above)
Comment (philei-tt) [reply tt-train/sources/ttml/ttml/autograd/function.py:47]: I feel like it's more explicit and gives more control, but not as nice 🤔

File: tt-train/sources/ttml/ttml/autograd/function.py (hunk 8 above)
Comment (ayerofieiev-tt) [reply tt-train/sources/ttml/ttml/autograd/function.py]: I agree, it is incredibly easy to mess-up.
My motivation was to better align with PyTorch. 
It is likely better to transfer weird knowledge from pytorch than to learn new thing.
No strong opinion here and I can revert to tt-train cpp style.

File: tt-train/sources/ttml/ttml/autograd/function.py (hunk 4 above)
Comment (ayerofieiev-tt) [reply tt-train/sources/ttml/ttml/autograd/function.py:47]: I am a lil worried that it allows to skip explicit capture of things in ctx

Comment (ayerofieiev-tt) [reply tt-train/sources/ttml/ttml/autograd/function.py:47]: if I understand it right, all variables within _call_ and implicitly enabled in the bw func in this case, right?

File: tt-train/sources/ttml/ttml/autograd/function.py (hunk 2 above)
Comment (ayerofieiev-tt) [reply tt-train/sources/ttml/ttml/autograd/function.py:104]: Both works. Let me leave as is, there is no real benefit to changing this.

File: tt-train/sources/ttml/ttml/autograd/function.py (hunk 3 above)
Comment (ayerofieiev-tt) [reply tt-train/sources/ttml/ttml/autograd/function.py]: Agreed. Will remove

File: tt-train/sources/ttml/ttml/autograd/function.py (hunk 4 above)
Comment (ayerofieiev-tt) [reply tt-train/sources/ttml/ttml/autograd/function.py:47]: I will keep 2 separate fwd/bw for now

File: tt-train/sources/ttml/ttml/autograd/__init__.py (hunk 5 above)
Comment (mdragulaTT) [reply tt-train/sources/ttml/ttml/autograd/__init__.py:24]: This is a bit off-topic and I know we have it somewhere back in our minds, but how hard would it be to unify ttml and ttnn Tensor types?

File: tt-train/sources/ttml/ttml/autograd/function.py (hunk 6 above)
Comment (mdragulaTT) [reply tt-train/sources/ttml/ttml/autograd/function.py]: Maybe I’m misunderstanding something here and/or how lora will work, but what about the case where a user doesn’t use lora at all? In that scenario this optimisation wouldn’t run, right? Is that desired?

File: tt-train/sources/ttml/ttml/autograd/__init__.py (hunk 5 above)
Comment (dyurshevichTT) [reply tt-train/sources/ttml/ttml/autograd/__init__.py:24]: I also prefer creating one node per operation, our c++ approach is good imo. Given that people using torch never define backward manually, we don't have to stick to that I guess

File: tt-train/sources/ttml/nanobind/nb_autograd.cpp [hunk 9]
```diff
@@ -208,12 +210,20 @@ void py_module(nb::module_& m) {
         py_auto_context.def("get_seed", &AutoContext::get_seed, "Get seed");
         py_auto_context.def(
             "add_backward_node",
-            &AutoContext::add_backward_node,
+            [](AutoContext& self, GradFunction grad_function, nb::object links_obj) {
```
Comment (athompsonTT) [inline tt-train/sources/ttml/nanobind/nb_autograd.cpp]: could links_obj be a nb::list?

**Review by athompsonTT — APPROVED** [review]

File: tt-train/sources/ttml/ttml/autograd/function.py (hunk 8 above)
Comment (ayerofieiev-tt) [reply tt-train/sources/ttml/ttml/autograd/function.py]: Actually after more digging and back and forth I decided to go with return, instead of add_grad

File: tt-train/sources/ttml/ttml/autograd/function.py (hunk 6 above)
Comment (philei-tt) [reply tt-train/sources/ttml/ttml/autograd/function.py]: This is generic optimization - we prune graph branches where all tensors have require_grad==False, it doesn't matter if that's lora or not

Comment (mdragulaTT) [reply tt-train/sources/ttml/ttml/autograd/function.py]: Ok, I was confused by this part:
> will be added as a part of lora pr

I remember the discussion with Denys about pruning and was just curious why it’s in the same PR as LoRA, but ok nvmd

**Review by philei-tt — APPROVED** [review]
